docs/source/cpp/env_vars.rst

```diff
@@ -148,19 +152,29 @@ that changing their value later will have an effect.
    platforms currently do not implement any form of runtime dispatch.
 
    .. note::
-      In addition to runtime dispatch, the compile-time SIMD level can
-      be set using the ``ARROW_SIMD_LEVEL`` CMake configuration variable.
-      Unlike runtime dispatch, compile-time SIMD optimizations cannot be
-      changed at runtime (for example, if you compile Arrow C++ with AVX512
-      enabled, the resulting binary will only run on AVX512-enabled CPUs).
-      Setting ``ARROW_USER_SIMD_LEVEL=NONE`` prevents the execution of
-      explicit SIMD optimization code, but it does not rule out the execution
-      of compiler generated SIMD instructions.  E.g., on x86_64 platform,
-      Arrow is built with ``ARROW_SIMD_LEVEL=SSE4_2`` by default.  Compiler
-      may generate SSE4.2 instructions from any C/C++ source code.  On legacy
-      x86_64 platforms do not support SSE4.2, Arrow binary may fail with
-      SIGILL (Illegal Instruction).  User must rebuild Arrow and PyArrow from
-      scratch by setting cmake option ``ARROW_SIMD_LEVEL=NONE``.
+      In addition to runtime dispatch, Arrow's build system supports two
+      compile-time CMake variables for controlling SIMD: ``ARROW_SIMD_LEVEL``
+      and ``ARROW_RUNTIME_SIMD_LEVEL``.
+
+      The ``ARROW_SIMD_LEVEL`` variable sets the minimum supported SIMD level at
+      the compiler level. This works by passing the appropriate target
+      architecture flag, such as ``-march=haswell`` for AVX2, to the compiler.
+      Unlike runtime dispatch, compile-time SIMD optimizations cannot be changed
+      at runtime (for example, if you compile Arrow C++ with AVX512 enabled, the
+      resulting binary will only run on AVX512-enabled CPUs).  Setting
+      ``ARROW_USER_SIMD_LEVEL=NONE`` prevents the execution of explicit SIMD
+      optimization code, but it does not rule out the execution of
+      compiler-generated SIMD instructions.  E.g., on x86, Arrow is built with
+      ``ARROW_SIMD_LEVEL=SSE4_2`` by default.  The compiler may generate SSE4.2
```

Comment:
Hmmm 🤔 A bit misleading here, since we've some hand-written SSE4.2 code, and some compiler generated code using some auto-vectorization techniques?

Reply:
I'm not familiar with either of those bits of the codebase but I can take a look tomorrow. Can you drop some links if you have them handy? (Logging off for the night here)

Reply:
https://github.com/apache/arrow/issues/40333#issuecomment-1976015427

I've pasted some in issue

Reply:
Thanks. It looks like, in those cases, the symbols get compiled in but, at least for the example in level_comparison.cc, the code looks like it runs through the dispatch machinery in dispatch.h so it still wouldn't be selected if `ARROW_USER_SIMD_LEVEL` is set in such a way that prevents it. All that code is new to me so I'd appreciate another set of eyes.

Reply:
No worries, and thank you for the help with the text here @mapleFU. It's appreciated.